bids/grabbids/config/derivatives.json

```diff
@@ -1,4 +1,6 @@
 {
+    "name": "derivatives",
+    "root": "derivatives",
```

Comment:
What does the `"root"` entry do? Derivatives can't be assumed to be in any particular location with respect to the BIDS root. So if this is saying "Apply this to a derivatives directory found in BIDS root", that's fine, but if it's saying that it must be found there to be applied, this is going to cause problems. (Also if it requires that "derivatives/" is in the path.)

Reply:
Yeah I wasn't sure how to best handle this. In theory, we could drop the `"root"` entry entirely, in which case it will fall back on the base BIDS directory. The thing I don't like about that is that the base BIDS spec currently excludes `./derivatives` (and some other folders) from indexing. The user is then able to index derivatives by specifying `config=['bids', 'derivatives']`, in which case the `derivatives.json` spec is invoked, which then removes the exclusion. If we were to remove the `"root"` entry, we probably would need to move the exclusion/inclusion to a new initialization argument that the user would have to specify much of the time, so that could get unwieldy.

It's also problematic because while separating directory structure from layout specification sounds good in principle, in practice, this means there would be no easy way to specify which configs apply to which directories--unless the user explicitly drops `layout.json` files into individual directories. My thinking was that it would be good to have defaults that handle the most common cases (i.e., where all the derivatives are in `derivatives/`), and then in cases where the derivatives are *not* where we expect them to be, users will then be responsible for making sure there are `layout.json` files where they're needed.

This seems like a decision that could have some design ramifications down the road, so maybe we should open it up for discussion. @chrisfilo, thoughts?

Reply:
There was a lot of push to make derivatives datasets standalone. I wonder if this might be the best angle here. The user might be required to point to the derivatives folder explicitly. Or in a more flexible way point to a set of folders which are assumed to include corresponding subjects and could be raw or derived data. For example

```
BIDSLayout('/my_dataset/', 
'/my_dataset/derivatives/freesurfer', 
'/my_dataset/derivatives/fmriprep', 
'/other_location/mriqc')
````

Reply:
@chrisfilo that's fine, I guess, but it leaves open the question of how to map derivatives folders onto layout configs. Just allowing the user to pass multiple folders wouldn't tell us which layouts apply to which folders. Do you see the `derivatives.json` config as only containing entity definitions that apply to *all* derivative directories? If so, then I think separating selection of layouts from specification of folders as a viable approach. The worry is that this could cause problems if people don't realize that adding `'derivatives'` to the `config` means that entities defined in `derivatives.json` will be matched *anywhere* in the search space--not just under `derivatives/`.

The new grabbit approach allows users to put very lightweight `layout.json` files in each directory, and those can specify which "domains" (i.e., configs) apply to the containing folder. So this wouldn't be too onerous, it just needs to be clearly explained. E.g., suppose you initialize:

```python
layout = BIDSLayout('/my_dataset', config=['bids', 'derivatives'], include=['/my_dataset/derivatives', '/other_location/mriqc'])`
```
(Technically that value for `config` is the default, so it could be left out). Suppose you don't want `derivatives.json` to apply to `/other_location/mriqc`. Then you could put a `layout.json` file in that directory that looks like this:

```json
{
  "domains": ["bids"]
}
```
...and then only `bids.json` would be applied.

Does that seem reasonable? An alternative would be to have a more complex initialization argument that can be a dict, with the keys being directories, and the values being the names of the domains that apply. E.g., `layout = BIDSLayout('/my_dataset', include={'/my_dataset/derivatives': ['bids', 'derivatives'], '/other_location/mriqc': ['bids']})`.

Reply:
I'm okay with any API that allows me to break with the default as a programmer (e.g. tagging a directory as a derivative or not), but requiring a `layout.json` in directories seems like it's bound to cause annoyance.

Reply:
i think @tyarkoni is saying that this allows one, but a user does not have to. but one can also see that something like a bids-app could generate this for the user and/or pybids can have a list of registered derivative layouts. how they will be versioned, etc.,. i don't know, but i think this does provide the flexibility.

@tyarkoni - i often don't like decoupled inputs that are actually related, so how about `BIDSLayout` accepting a list of dictionaries along the following lines:

```
[{ "locations": ['/my_dataset', '/my_dataset/derivatives'],
   "config": ['bids', 'derivatives']
},
{ "locations": ['/my_dataset/derivatives/freesurfer'],
   "config": ['bids', 'freesurfer']
},
]
``

Reply:
Sure, BIDS apps can do that, but the JSON config is a grabbit convention, and is ostensibly what's needed to make the software conform to the spec. This introduction of a new feature and making it required is bypassing the spec process.

I'm not trying to be a rules lawyer here, but we have derivatives in the wild that for obvious reasons lack this config file, and until it's in the spec, we can't expect any non-pybids-generated derivatives to include it.

Reply:
pybids doesn't have to respect grabbit conventions in the user-facing API; I'm happy to hide all that stuff internally if needed. The question I'm concerned with is what the user-facing API should look like. I have some ideas though, will implement something and push for review.

Reply:
Question for @effigies @chrisfilo @satra: do you think that by default, a `BIDSLayout` should omit `derivatives/`, `models/`, `code/`, and such from indexing? That's the current behavior, but on reflection, I'm not sure this makes sense. It feels to me like the default should be to look *everywhere* under the project root for filenames containing valid BIDS entities, not just within the subject hierarchy.

One reason this is important to get right is that, as once we start adding optional configs the user can  name, (e.g., `layout = BIDSLayout('.', config=['bids', 'derivatives', 'models'])`), it becomes ambiguous as to what the intended indexing behavior is. Should files in `/derivatives` directories automatically be indexed for entities defined in *both* `derivatives.json` or `bids.json`? I lean towards the former, and then the user can specify the `exclude` directive for the `'bids'` config if they need to. The alternative (the user has to explicitly specify that they want BIDS to apply to the domains of other configs) seems too clunky.

Reply:
> do you think that by default, a BIDSLayout should omit derivatives/, models/, code/, and such from indexing?

No, I think any subdirectory should be included. Is there a good argument against it? What was the justification for not doing so before? Just that the specs for those directories were undefined?

> One reason this is important to get right is that, as once we start adding optional configs the user can name, (e.g., layout = BIDSLayout('.', config=['bids', 'derivatives', 'models'])), it becomes ambiguous as to what the intended indexing behavior is.

This may already be ambiguous. That says to me that the current directory is a derivative that may contain models. Are you expecting it to say that you expect `./derivatives/` and `./models/` subdirectories?

> Should files in /derivatives directories automatically be indexed for entities defined in both derivatives.json and bids.json

Yes.

> The alternative (the user has to explicitly specify that they want BIDS to apply to the domains of other configs) seems too clunky.

I agree. BIDS entities will usually be relevant, so the default should be to include them.

-----

At the risk of being redundant: My concern in all of this has not been one of calling convention, but that an isolated derivative might not be indexed the same as one that is in `<bidsroot>/derivatives/`.

Reply:
Cool, I think we're on the same page. The changes I'm about to push will allow users to easily map any named config to any directory (per my earlier suggestion, but slightly more flexible), so that shouldn't be a problem. By default, including the `'derivatives'` config by name will only cause it to look in `'derivatives/'`, but the user can easily do something like:

```
layout = BIDSLayout('.', config=['bids', ('derivatives', ['/derivatives', 'more_derivatives'])])
```

The user can also still inject `layout.json` files into specific directories and they will be handled per grabbit conventions, but this won't be documented anywhere in pybids, and is not necessary for anything to work.

Reply:
Oh, and the original concern about indexing was just about indexing speed: `derivatives/` folders could conceivably be very large, and since grabbit isn't optimized for speed, indexing everything by default could easily add a couple of minutes to initialization for large projects. But I guess my feeling about this is that if that happens, the user is free to manually exclude paths.

Reply:
Okay. When pushed, I'll give your branch a shot and make sure I can do what I need.

Also, I'm currently [doing the following](https://github.com/poldracklab/fitlins/blob/b8a4a89ff19ce466443329eb7d5c1ca5fc4decf9/fitlins/base.py#L206-L210):

```Python
    fl_layout = grabbids.BIDSLayout(
        deriv_dir,
        extensions=['derivatives',
                    pkgr.resource_filename('fitlins', 'data/fitlins.json')])
```

If I understand you correctly, I should do this:

```Python
fitlins_json_path = pkgr.resource_filename('fitlins', 'data/fitlins.json')
fl_layout = BIDSLayout(deriv_dir, config=['bids', ('derivatives', ['.']),
                                          (fitlins_json_path, ['.']))
```

Reply:
I think that's right, and should work. Note that you'll also be able to pass a full config dictionary as the second tuple element (which overwrites the dict extracted from the first argument), so that's how one can easily overwrite things like `include` and `exclude` if needed).

Reply:
Okay I think it's basically ready for release, but I'll hold off merging and releasing until you can try it out and see if there's anything that needs to be fixed.